internal/transport/http2_client.go

```diff
@@ -702,7 +707,18 @@ func (e NewStreamError) Error() string {
 // streams.  All non-nil errors returned will be *NewStreamError.
 func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (*Stream, error) {
 	ctx = peer.NewContext(ctx, t.getPeer())
-	headerFields, err := t.createHeaderFields(ctx, callHdr)
+
+	dupCallHdr := *callHdr
+	// ServerName field of the resolver returned address takes precedence over
+	// Host field of CallHdr to determine the :authority header. This is because,
+	// the ServerName field takes precedence for server authentication during
+	// TLS handshake, and the :authority header should match the value used
+	// for server authentication.
+	if t.address.ServerName != "" {
+		dupCallHdr.Host = t.address.ServerName
+	}
```

Comment:
@dfawley : I asked @holdno to make this copy here, because I wasn't completely sure of the downstream effects of changing the `CallHdr.Host` field. Could you please take a look. Thanks.

Reply:
I think that's the right call; otherwise if you reuse `callHdr` across multiple attempts for one RPC, and one transport has an authority set and another doesn't, then you could end up with the wrong behavior.

Can we avoid making this copy, though, if there is no change?

```go
if t.address.ServerName != "" {
	newCallHdr := *callHdr
	newCallHdr.Host = t.address.ServerName
	callHdr = &newCallHdr
}
```

Reply:
thanks for your advice, it looks better.